xds/src/main/java/io/grpc/xds/GrpcAuthorizationEngine.java

```diff
@@ -0,0 +1,381 @@
+      this.serverCall = serverCall;
+    }
+
+    private String getPath() {
+      return serverCall.getMethodDescriptor().getFullMethodName();
+    }
+
+    private Collection<String> getPrincipalNames() {
+      SSLSession sslSession = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_SSL_SESSION);
+      Set<String> principalNames = new HashSet<>();
+      try {
+        Certificate[] certs = sslSession == null ? null : sslSession.getPeerCertificates();
+        if (certs == null || certs.length < 1) {
+          return principalNames;
+        }
+        X509Certificate cert = (X509Certificate)certs[0];
+        if (cert == null || cert.getSubjectDN() == null) {
+          return principalNames;
+        }
+        Collection<List<?>> names = cert.getSubjectAlternativeNames();
+        if (names != null) {
+          for (List<?> name : names) {
+            if (URI_SAN == (Integer) name.get(0)) {
+              principalNames.add((String) name.get(1));
+            }
+          }
+          if (!principalNames.isEmpty()) {
+            return principalNames;
+          }
+          for (List<?> name : names) {
+            if (DNS_SAN == (Integer) name.get(0)) {
+              principalNames.add((String) name.get(1));
+            }
+          }
+          if (!principalNames.isEmpty()) {
+            return principalNames;
+          }
+        }
+        return Collections.singleton(cert.getSubjectDN().getName());
+      } catch (SSLPeerUnverifiedException | CertificateParsingException ex) {
+        return null;
+      }
+    }
+
+    private Metadata getHeader() {
+      return metadata;
+    }
+
+    private InetAddress getDestinationIp() {
+      SocketAddress addr = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_LOCAL_ADDR);
+      return addr == null ? null : ((InetSocketAddress) addr).getAddress();
+    }
+
+    private InetAddress getSourceIp() {
+      SocketAddress addr = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
+      return addr == null ? null : ((InetSocketAddress) addr).getAddress();
+    }
+
+    private int getDestinationPort() {
+      SocketAddress addr = serverCall.getAttributes().get(Grpc.TRANSPORT_ATTR_LOCAL_ADDR);
+      return addr == null ? -1 : ((InetSocketAddress) addr).getPort();
+    }
+  }
+
+  abstract static class Matcher {
+    protected Matcher() {
+    }
+
+    abstract boolean matches(EvaluateArgs args);
+  }
+
+  /** Matches when any of the matcher matches. */
+  static final class OrMatcher extends Matcher {
+    private final List<? extends Matcher> anyMatch;
+
+    OrMatcher(List<? extends Matcher> matchers) {
+      checkNotNull(matchers, "matchers");
+      this.anyMatch = Collections.unmodifiableList(matchers);
+    }
+
+    static OrMatcher create(Matcher...matchers) {
+      return new OrMatcher(Arrays.asList(matchers));
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      for (Matcher m : anyMatch) {
+        if (m != null && m.matches(args)) {
```

Comment:
We should throw in the constructor if any entry is `null` instead of worrying about it here. It doesn't seem there is any value is allowing null. Ditto for AndMatcher.